Mob AI adjustments

MuckerMayhem opened the pull request:
:cl: Mucker
tweak: Made it more difficult to dodge simple mobs melee attacks.
tweak: Mobs calling for help will no longer cascade and bring the entire map toward them.
/:cl:

Mobs targetting is now based on direction. If a target is adjacent to the mob as it melees, and the target is in the direction the mob is facing, then it will count as a hit, despite movement from the target.

Mobs that are called upon for help will no longer immediately call out for help as well, and the range at which the call for help is heard was reduced, which seems to fix the problem of mobs coming from all corners when one called for help.

SierraKomodo commented on code/modules/ai/ai_holder_cooperation.dm:
```diff
@@ -1,8 +1,9 @@
-	var/call_distance = 14						// How far away calls for help will go for.
+	var/call_distance = 10						// How far away calls for help will go for.
 	var/last_helpask_time = 0					// world.time when a mob asked for help.
+	var/last_call_time = 0                      // world.time when mob was called upon for help.
```
```suggestion
	/// `world.time` when mob was called upon for help.
	var/last_call_time = 0
```

SierraKomodo commented on code/modules/ai/ai_holder_cooperation.dm:
```diff
@@ -73,7 +74,7 @@
 	ai_log("request_help() : Exiting.", AI_LOG_DEBUG)
 
-// What allies receive when someone else is calling for help.1
+// What allies receive when someone else is calling for help.
```
```suggestion
/// What allies receive when someone else is calling for help.
```

SierraKomodo replied:
Still isn't a dmdoc :P

SierraKomodo commented on code/modules/ai/ai_holder_cooperation.dm:
```diff
@@ -1,17 +1,18 @@
+	var/last_helpask_time = 0
+	/// world.time when mob was called upon for help.
+	var/last_call_time = 0
+	/// List of all mobs inside the faction with ai_holders that have cooperate on, to call for help without using range().
+	var/list/faction_friends = list()
 												// Note that this is only used for sending calls out. Receiving calls doesn't care about this list, only if the mob is in the faction.
 												// This means the AI could respond to a player's call for help, if a way to do so was implemented.
```
These last two comments look like they're meant to be part of the `faction_friends` documentation. Could also make it multiline but the dm-specific syntax for that escapes me at the moment.
```suggestion
	/// List of all mobs inside the faction with ai_holders that have cooperate on, to call for help without using range(). Note that this is only used for sending calls out. Receiving calls doesn't care about this list, only if the mob is in the faction. This means the AI could respond to a player's call for help, if a way to do so was implemented.
	var/list/faction_friends = list()
```

Spookerton commented on code/modules/ai/ai_holder_cooperation.dm:
```diff
@@ -1,16 +1,15 @@
-	// It was probably worth it.
-	var/call_players = FALSE					// (Currently nonfunctional) If true, players get notified of an allied mob calling for help.
-	var/called_player_message = "needs help!"	// (Currently nonfunctional) Part of a message used when above var is true. Full message is "\The [holder] [called_player_message]"
+	/// If true, asks allies to help when fighting something.
+	var/cooperative = FALSE
+	/// How far away calls for help will go for.
+	var/call_distance = 10
+	/// world.time when a mob asked for help.
+	var/last_helpask_time = 0
+	/// world.time when mob was called upon for help.
+	var/last_call_time = 0
+	/// List of all mobs inside the faction with ai_holders that have cooperate on, to call for help without using range(). Note that this is only used for sending calls out. Receiving calls doesn't care about this list, only if the mob is in the faction.
+	var/list/faction_friends = list()
```
It's generally best to throw in newlines for (visual) context control if you're going to do pre comments instead of same line

Spookerton commented on code/modules/ai/ai_holder_cooperation.dm:
```diff
@@ -1,16 +1,15 @@
+	/// How far away calls for help will go for.
+	var/call_distance = 10
+	/// world.time when a mob asked for help.
+	var/last_helpask_time = 0
+	/// world.time when mob was called upon for help.
+	var/last_call_time = 0
```
prefer clearer names for these two vars; `last_received_help_request` and `last_sent_help_request` or something for their current use.

would also be beneficial (in a tiny way) to invert them like
```
var/next_receive_help_request
var/next_send_help_request
...
if (!cooperative || world.time < next_receive_help_request || world.time < next_send_help_request)
    return
...
next_receive_help_request = world.time + 15 SECONDS
...
next_send_help_request = world.time + 10 SECONDS
```

MuckerMayhem committed:
tweak mob attack attack 'aim'

AI attacks now check for hits based on direction. If a target is adjacent to the mob and in the direction it is facing, then it will count as a hit.

MuckerMayhem committed:
nerf mob AI call for help mechanic
